lib/wasi/src/syscalls/mod.rs

```diff
@@ -76,16 +75,18 @@ fn write_bytes<T: Write>(
 fn read_bytes<T: Read>(
     mut reader: T,
     memory: &Memory,
-    iovs_arr_cell: &[Cell<__wasi_iovec_t>],
+    iovs_arr_cell: Vec<WasmCell<__wasi_iovec_t>>,
 ) -> Result<u32, __wasi_errno_t> {
     let mut bytes_read = 0;
 
     for iov in iovs_arr_cell {
         let iov_inner = iov.get();
-        let bytes = WasmPtr::<u8, Array>::new(iov_inner.buf).deref(memory, 0, iov_inner.buf_len)?;
-        let mut raw_bytes: &mut [u8] =
-            unsafe { &mut *(bytes as *const [_] as *mut [_] as *mut [u8]) };
-        bytes_read += reader.read(raw_bytes).map_err(|_| __WASI_EIO)? as u32;
+        let mut raw_bytes: Vec<u8> = vec![0; iov_inner.buf_len as usize];
```

Comment:
To slightly mitigate the extra slowdown we can move the vector outside the loop, call `raw_bytes.clear()` on every loop to reset it and then `raw_bytes.resize(iov_inner.buf_len as usize, 0)`  this will prevent reallocation over and over again. Note that `resize` never shrinks the underlying allocation so it'll just make it bigger if needs to.

Though honestly this is a prime candidate for optimization, I think this code is likely the hardest hit by these API changes.

Reply:
Good call!

Reply:
Done